---
title: "000-PREVOT-HME — PREPROCESSING"
author: "Mbouamboua Yvon <yvon.mbouamboua@inserm.fr>"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
date: '`r format(Sys.Date())`'
---

## Goal
Quality control and normalization

## Setup
```{r setup, include=TRUE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, comment = "")
dir.main <- "/Users/yvon.mbouamboua/projects/singlecell/000-prevot-hme"
dir.results <- file.path(dir.main, "output", "000-sauve-2025")
dir.create(dir.results, showWarnings = FALSE, recursive = TRUE)

# Load packages with automatic installation if missing
packages <- c(
  "Seurat", "ggplot2", "dplyr", "grid",
  "SingleCellExperiment", "scran", "scater",
  "pheatmap", "ComplexHeatmap", "ComplexUpset", "patchwork")
install_and_load <- function(pkg) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
    BiocManager::install(pkg, ask = FALSE)
  }
  suppressWarnings(suppressMessages(library(pkg, character.only = TRUE)))
}
suppressPackageStartupMessages({
  invisible(lapply(packages, install_and_load))
})

# Load helper functions
utils_dir <- "~/projects/apps/"
r_files <- list.files(utils_dir, pattern = "\\.R$", full.names = TRUE)
invisible(lapply(r_files, source))
```


# Create seurat objects

```{r create_seurat_object}

file_name <- list("AD_23425","AD_29405", "CTRL_26450", "CTRL_24457")

for (i in 1:length(file_name)) {
  print(file_name[[i]])  
  counts <- Read10X_h5(file.path(dir.data, file_name[[i]], h5))
  sobj <- CreateSeuratObject(counts = counts, project = file_name[[i]])
  saveRDS(sobj, file.path(dir.results, "rdata", filename = paste0(file_name[[i]], ".rds")))
  rm(sobj)
}

```

# Load seurat objects

```{r}

seurat_list <- list()
file_name <- list("AD_23425","AD_29405", "CTRL_26450", "CTRL_24457")

for (i in 1:length(file_name)) {
  file_path = list.files(path = file.path(dir.results, "rdata"), pattern =  paste0(file_name[[i]], ".rds"), 
                         full.names = T, recursive = F)
  sobj <- readRDS(file_path)
  print(unique(sobj$orig.ident))
  seurat_list[[file_name[[i]]]] <- sobj
  rm(sobj)
}

```


# Filter quality

```{r filter_quality}

merged <- Run_Quality_Control(seurat_input = seurat_list, sample_column = "orig.ident", min_features = 200, max_mito = 5, 
                           filter_method = "MAD", MAD = 10, calculate_dropout = F, filter_max_counts = T,filter_doublets = T,
                           merge_output = T,dir_results = file.path(dir.results, "QC"))

saveRDS(merged, file.path(dir.main, "output/000-preprocessing/data/merged_2AD_2CTRL.rds"))

```


# Harmony integration

```{r fig.width=8, fig.height=6}

obj <- readRDS(file.path(dir.main, "output/000-preprocessing/data/merged_2AD_2CTRL.rds"))

obj <- NormalizeData(obj)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- RunPCA(obj, npcs = 50)
obj <- FindNeighbors(obj, dims = 1:50, reduction = "pca")
obj <- FindClusters(obj, resolution = 1, cluster.name = "unintegrated_clusters")
obj <- IntegrateLayers(obj, method = HarmonyIntegration, orig.reduction = "pca", new.reduction = "harmony",verbose = FALSE)
obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:50)
obj <- FindClusters(obj, resolution = 1, cluster.name = "harmony_clusters")
obj <- RunUMAP(obj, reduction = "harmony", dims = 1:50, reduction.name = "umap", min.dist = 0.5)
obj <- JoinLayers(obj)
Dim_Plot(object = obj, group.by = "seurat_clusters", label.show = T, pt.size = 0.5, legend = T, dark.theme = F, figure.plot = T)


```


# Remove low quality clusters (probabily the doublets)

```{r fig.width=8, fig.height=6}

Idents(obj) <- "seurat_clusters"
obj <- subset(obj, idents = c(8,10,23,28,29,33,34), invert = T)
obj <- NormalizeData(obj)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- harmony::RunHarmony(obj, group.by.vars = "orig.ident")
obj <- FindNeighbors(obj, dims = 1:50, reduction = "harmony")
obj <- FindClusters(obj, resolution = 1.5, cluster.name = "seurat_clusters")
ndims <- Get_PCs(obj, reduction.name = "harmony")
obj <- RunUMAP(obj, dims = 1:ndims, reduction = "harmony", reduction.name = "umap", min.dist = 1, spread = 2)
Dim_Plot(obj, group.by = "seurat_clusters", label.show = T, legend = F, dark.theme = T, figure.plot = T)

saveRDS(obj, file.path(dir.main, "output/000-integrated/data/integrated.rds"))

```


# Load integrated data

```{r}

obj <- readRDS(file.path(dir.main, "output/000-integrated/data/integrated.rds"))

```


# Find markers

```{r fig.width=8, fig.height=25}

DefaultAssay(obj) <- "RNA"
sub <- Filter_Unwanted_Genes(obj, scale_data = FALSE)
markers <- Find_Markers(sub,  group.by = "seurat_clusters", only.pos = T, max.cells.per.ident = 1000)
rm(sub)
write.table(markers, file.path(dir.results, "tables/seurat_markers.tsv"), sep = "\t", row.names = F, quote = F)
markers <- read.delim(file.path(dir.results, "tables/seurat_markers.tsv"))
top <- markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
View(top)
Idents(obj) <- "seurat_clusters"
invisible(gc())
p <- Dot_Plot(obj, features = rev(top$gene), flip.axes = T, dot.scale = 8, base.size = 10, theme.bw = F, col.min = 0, angle.x = 90)
p
save_plot(p, filename = file.path(dir.results, "figures/seurat_markers"),formats = c("pdf"),  width = 20, height = 8)    

```


```{r}

dim_plot(obj, group.by = "seurat_clusters", label = T, legend = F, dark = T, fig.plot = T)

```


# Manual annotation

```{r fig.width=14, fig.height=14}

# Neuron
Neuron <-  c("SLC32A1","GAD1","GAD2", # GABA (Xin Zhou, 2022)
"SLC17A6","SLC17A7","CRH", # GLUT (Xin Zhou, 2022)
"SLC18A2","HDC", # Histaminergic (Xin Zhou, 2022)
"AVP","TAC3","GAL","KISS1","OXT","POMC","SIM1","NPY","COL12A1","TH","NOS1","NEUROD4"
)

Feature_Plot(obj, features = Neuron, no_axes = T)
Stacked_VlnPlot(obj, group.by="seurat_clusters", features = c("SLC32A1","GAD1","GAD2","SLC17A6","SLC17A7"), pt.size = 0) 

## Peri
Feature_Plot(obj, features = "PDGFRB", no_axes = T)
pericyte.cells <- get_gene_cells(obj, clusters = c(22), cluster_col = "seurat_clusters", genes =  c("PDGFRB"), threshold = 2)

## VLMC
Feature_Plot(obj, features = "BMP5", no_axes = T)

## Fenestrated EC
Feature_Plot(obj, features = "Fenestrated-EC", no_axes = T)
plvap.cells <- get_gene_cells(obj, clusters = c(22), cluster_col = "seurat_clusters", genes =  c("PLVAP", "HSPB1", "EHD4","TCIM"), threshold = 2)

Idents(obj) <- "seurat_clusters"
#obj <- FindSubCluster(obj, cluster = c(24),subcluster.name = "C24", graph.name = "RNA_snn", resolution = 0.2)
#dim_plot(object = obj, group.by = "C24", label = T)
obj <- FindSubCluster(obj, cluster = c(35), subcluster.name = "C35",graph.name = "RNA_snn", resolution = 0.1)
dim_plot(object = obj, group.by = "C35", label = T)

Idents(obj) <- "seurat_clusters"
sub <- Seurat_Subset(obj, cluster_id = 24, ndims = 4, resolution = 0.1)
dim_plot(object = sub, group.by = "seurat_clusters")
fenestrated_EC <- c("VWF","CLDN", "Fenestrated-EC","EXOC3L2", "HSPB1", "EHD4","TCIM")
Feature_Plot(sub, features = fenestrated_EC, na_cutoff = 1.5, no_axes = T)
#plvap.cells <- WhichCells_Gene_Expressions(obj, group.by = "seurat_clusters", idents = 24, features = c("Fenestrated-EC", "HSPB1", "EHD4","TCIM"), min.expr = 2)
#plvap.cells <- WhichCells(sub, idents = 1)

## Astro 
astro <- c("AQP4","GFAP", "SLC13A5","SLC7A10","KISS1","OXT","DIO2")
Dot_Plot(obj, features = astro, flip.axes = T, angle.x = 90, plot.title = "Astrocytes")
Feature_Plot(obj, features = astro, no_axes = T)

## Ependy 
ependy <- c("FOXJ1","STOML3","CCDC153","PIFO", "DYNLRB2")
Dot_Plot(obj, features = ependy, flip.axes = T, angle.x = 90, plot.title = "Ependymocytes")
Feature_Plot(obj, features = ependy, no_axes = T, ncol = 4)

## A1 tany
A1 <- c("C2CD4A","HAS2","SLC14A2","CCL2","SHISA9","SYNPR","CXCL2","SLC1A2","CHST8","CXCL12","VIM","DIO2")
Dot_Plot(obj, features = A1, flip.axes = T, angle.x = 90, plot.title = "A1 Tany")
Feature_Plot(obj, features = c(A1,"AGT","TGFB2"), no_axes = T)
Feature_Plot(obj, features = c(A2, "SLC7A11","EPHB1"), no_axes = T)
Feature_Plot(obj, features = c("DIO2","AQP4","S100A10","GPX3","SCG3","TPPP3","SV2C","CD44"), no_axes = T)
a2.cells <- WhichCells(obj, idents = 1)

## Tany A2
A2 <- c("FSHR","LTF","LRAT","CCDC85A","PRR16","GALNT15","ID3","CD44","DUSP2","PLCXD3","VCAN","DPP10","NRXN3","COL24A1")
Dot_Plot(obj, features = A2, flip.axes = T, angle.x = 90, plot.title = "A2 Tany")
Feature_Plot(obj, features = A2, no_axes = T)

## Tany-interferon
tany_inf <- c("GPX3","IFITM3","IFI27","MT1X","CXCL14","SFRP2","KDR","IGFBP5","TIMP1","IRF7")
Dot_Plot(obj, features = tany_inf, flip.axes = T, angle.x = 90, plot.title = "Tany-interferon")
Feature_Plot(obj, features = tany_inf, no_axes = T)

## B1 tany
B1 <- c("CRYM","FRZB","NPSR1","TPH1","NRG1","DIO2","FGF10","CXCL3","COL4A3","CUX2")
Dot_Plot(obj, features = B1, flip.axes = T, angle.x = 90, plot.title = "B1 Tany")
Feature_Plot(obj, features = B1, no_axes = T, ncol = 4)

## B2
B2 <- c("COL25A1","FGF10","A2M","MEST","FRZB","PENK","CRYM")
Dot_Plot(obj, features = B2, flip.axes = T, angle.x = 90, plot.title = "B2 Tany")
Feature_Plot(obj, features = B2, no_axes = T, ncol = 4)

## Immune
Feature_Plot(obj, features = c("THEMIS","PTPRC","MARCO"), no_axes = T, merge_legend = T)

## Micg annotation
Feature_Plot(obj, features = c("CSF1R", "MERTK", "CX3CR1"), no_axes = T, na_cutoff = 1, ncol = 3)

## MAC
Feature_Plot(obj, features = c("MARCO","CD68"), no_axes = T)
```

## ann1

```{r}

obj@meta.data$ann1 <- NULL
obj@meta.data[obj@meta.data$seurat_clusters %in% c(9,27,39,4,18,19,23,36,11,33,17,24,34,28,38,37,32,35,31,15,14,16,24,26,29),"ann1"] <-"Neurons"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(5,20,10),"ann1"] <-"Tanycytes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(1,7,12),"ann1"] <-"Astrocytes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(0,6,8,40),"ann1"] <-"Immunes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(25),"ann1"] <-"Ependymocytes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(2,3,13,21),"ann1"] <-"Oligodendrocytes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(22,30),"ann1"] <-"Vasculars"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(41),"ann1"] <-"Myelinating-glial"

dim_plot(obj, group.by = "ann1", label = T, legend = T, dark = T, fig.plot = T, 
         label.fontface = "italic", label.size = 3.5,  pt.size = 0.5)

```


## ann2

```{r}

obj@meta.data$ann2 <- NULL
obj@meta.data[obj@meta.data$seurat_clusters %in% c(9,27,39),"ann2"] <-"Glutamatergic"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(4,18,19,23,36,11,33,17,24,34,28,38,37,32,35,31,15,14,16,24,26,29),"ann2"] <-"GABAergic"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(5,10,20),"ann2"] <-"Tanycytes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(1,7,12),"ann2"] <-"Astrocytes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(6),"ann2"] <-"Myeloid"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(0,8,40),"ann2"] <-"Microglia"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(25),"ann2"] <-"Ependymocytes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(3),"ann2"] <-"Oligo-precursors"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(2,13,21),"ann2"] <-"Oligodendrocytes"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(22),"ann2"] <-"Endothelial"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(41),"ann2"] <-"Myelinating-glial"
obj@meta.data[rownames(obj@meta.data) %in% pericyte.cells,"ann2"] <- "Pericytes" # PDGFRB
obj@meta.data[obj@meta.data$seurat_clusters %in% c(30),"ann2"] <-"VLMCs"

dim_plot(obj, group.by = "ann2", label = T, legend = T, dark = T, fig.plot = T, 
         label.fontface = "italic", label.size = 3.5, pt.size = 0.5)

```


## ann3

```{r}

obj@meta.data$ann3 <- NULL
obj@meta.data[obj@meta.data$seurat_clusters %in% c(9,27,39),"ann3"] <-"Glutamatergic"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(4,18,19,23,36,11,33,17,24,34,28,38,37,32,35,31,15,14,16,24,26,29),"ann3"] <-"GABAergic"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(1,7,12),"ann3"] <-"Astrocytes" # AGT, AQP4
obj@meta.data[obj@meta.data$seurat_clusters %in% c(10),"ann3"] <-"α1-tanycytes" # SLC14A2
#obj@meta.data[obj@meta.data$seurat_clusters %in% c(),"ann3"] <-"A2-tanycytes" # LRAT
obj@meta.data[obj@meta.data$seurat_clusters %in% c(5),"ann3"] <-"α2-β1-tanycytes" # CRYM, FRZB
obj@meta.data[obj@meta.data$seurat_clusters %in% c(20),"ann3"] <-"β2-tanycytes" # COL25A1
#obj@meta.data[obj@meta.data$seurat_clusters %in% c(),"ann3"] <-"Tany-interferon" # IFITM3
obj@meta.data[obj@meta.data$seurat_clusters %in% c(25),"ann3"] <- "Ependymocytes" # CCDC153
obj@meta.data[obj@meta.data$seurat_clusters %in% c(6),"ann3"] <-"Macrophages" # MARCO
obj@meta.data[obj@meta.data$seurat_clusters %in% c(0,8,40),"ann3"] <- "Microglia" # CX3CR1
obj@meta.data[obj@meta.data$seurat_clusters %in% c(3),"ann3"] <-"Oligo-precursors" # OLIG1
obj@meta.data[obj@meta.data$seurat_clusters %in% c(2,13,21),"ann3"] <-"Oligodendrocytes" # MBP
obj@meta.data[obj@meta.data$seurat_clusters %in% c(22),"ann3"] <-"Endothelial" # VWF
obj@meta.data[rownames(obj@meta.data) %in% plvap.cells,"ann3"] <- "Fenestrated-EC" # PLVAP
obj@meta.data[rownames(obj@meta.data) %in% pericyte.cells,"ann3"] <- "Pericytes" # PDGFRB
obj@meta.data[obj@meta.data$seurat_clusters %in% c(30),"ann3"] <-"VLMCs" # BMP5
obj@meta.data[obj@meta.data$seurat_clusters %in% c(41),"ann3"] <-"Myelinating-glial" # CDH19

dim_plot(obj, group.by = "ann3", label = T, legend = T, dark = T, fig.plot = T, 
         label.fontface = "italic", label.size = 3.5,pt.size = 0.5)

```

```{r}

cellmap(hme_final, group.by = "ann3", label = T)
cellmap(hme_final, group.by = "ann2", label = T)
cellmap(hme_final, group.by = "ann1", label = T)
cellmap(hme_final, group.by = "ann4", label = T)

names(hme_final@meta.data)

```


## ann4

```{r fig.width=10, fig.height=10}

obj@meta.data$ann4 <- NULL
obj@meta.data[obj@meta.data$seurat_clusters %in% c(1,7,12),"ann4"] <-"Astrocytes" # AGT, AQP4
obj@meta.data[obj@meta.data$seurat_clusters %in% c(10),"ann4"] <-"α1-tanycytes" # SLC14A2
obj@meta.data[obj@meta.data$seurat_clusters %in% c(5),"ann4"] <-"α2-β1-tanycytes" # CRYM, FRZB
obj@meta.data[obj@meta.data$seurat_clusters %in% c(20),"ann4"] <-"β2-tanycytes" # COL25A1
obj@meta.data[obj@meta.data$seurat_clusters %in% c(25),"ann4"] <- "Ependymocytes" # CCDC153
obj@meta.data[obj@meta.data$seurat_clusters %in% c(6),"ann4"] <-"Macrophages" # MARCO
obj@meta.data[obj@meta.data$seurat_clusters %in% c(0,8,40),"ann4"] <- "Microglia" # CX3CR1
obj@meta.data[obj@meta.data$seurat_clusters %in% c(3),"ann4"] <-"Oligo-precursors" # OLIG1
obj@meta.data[obj@meta.data$seurat_clusters %in% c(2,13,21),"ann4"] <-"Oligodendrocytes" # MBP
obj@meta.data[obj@meta.data$seurat_clusters %in% c(22),"ann4"] <-"Endothelial" # VWF
obj@meta.data[rownames(obj@meta.data) %in% plvap.cells,"ann4"] <- "Fenestrated-EC" # PLVAP
obj@meta.data[rownames(obj@meta.data) %in% pericyte.cells,"ann4"] <- "Pericytes" # PDGFRB
obj@meta.data[obj@meta.data$seurat_clusters %in% c(30),"ann4"] <-"VLMCs" # BMP5
obj@meta.data[obj@meta.data$seurat_clusters %in% c(41),"ann4"] <-"Myelinating-glial" # CDH19
obj@meta.data[obj@meta.data$seurat_clusters %in% c(36),"ann4"] <-"AVP+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(16),"ann4"] <-"POMC+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(11,23),"ann4"] <-"KISS1+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(15),"ann4"] <-"NPY+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(28),"ann4"] <-"AGRP+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(27),"ann4"] <-"ONECUT1+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(26),"ann4"] <-"GRIP2+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(14),"ann4"] <-"CYSLTR2+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(29),"ann4"] <-"NDST4+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(32),"ann4"] <-"INPP4B+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(35),"ann4"] <-"LHX1+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(34),"ann4"] <-"RELN+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(4),"ann4"] <-"CALM3+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(17),"ann4"] <-"GAL+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(18),"ann4"] <-"GAL+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(33),"ann4"] <-"ZNF831+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(37),"ann4"] <-"SLC26A4+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(38),"ann4"] <-"HDC+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(9),"ann4"] <-"CHRM2+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(39),"ann4"] <-"LMX1A+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(24),"ann4"] <-"MTUS2+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(31),"ann4"] <-"THSD7B+"
obj@meta.data[obj@meta.data$seurat_clusters %in% c(19),"ann4"] <-"GALNTL6+"

dim_plot(obj, group.by = "ann4", label = T, legend = T, dark = T, fig.plot = T, 
         label.fontface = "italic", label.size = 3.5, pt.size = 0.5)

```


# Add metadata

```{r}

obj@meta.data$group <- NULL
obj@meta.data[obj@meta.data$orig.ident %in% c("AD_23425","AD_29405"),"group"] <-"AD" 
obj@meta.data[obj@meta.data$orig.ident %in% c("CTRL_26450","CTRL_24457"),"group"] <-"CTRL" 
obj$group <- factor(obj$group, levels = c("AD","CTRL"))

obj@meta.data$sex <- NULL
obj@meta.data[obj@meta.data$orig.ident %in% c("CTRL_26450"),"sex"] <-"F"
obj@meta.data[obj@meta.data$orig.ident %in% c("AD_23425","AD_29405","CTRL_24457"),"sex"] <-"M"

obj@meta.data$age <- NULL
obj@meta.data[obj@meta.data$orig.ident %in% c("AD_29405"),"age"] <-"74" 
obj@meta.data[obj@meta.data$orig.ident %in% c("AD_23425"),"age"] <-"61" 
obj@meta.data[obj@meta.data$orig.ident %in% c("CTRL_24457"),"age"] <-"71" 
obj@meta.data[obj@meta.data$orig.ident %in% c("CTRL_26450"),"age"] <-"63" 

obj@meta.data$sample <- NULL
obj@meta.data[obj@meta.data$orig.ident %in% "AD_23425","sample"] <- "AD-23425"
obj@meta.data[obj@meta.data$orig.ident %in% "AD_29405","sample"] <- "AD-29405"
obj@meta.data[obj@meta.data$orig.ident %in% "CTRL_26450","sample"] <- "CTRL-26450"
obj@meta.data[obj@meta.data$orig.ident %in% "CTRL_24457","sample"] <- "CTRL-24457"

obj$sample <- factor(obj$sample, levels = c("AD-29405","AD-23425","CTRL-26450","CTRL-24457"))

```



## Order the cells

```{r}

obj@meta.data$ann1 <- factor(obj@meta.data$ann1, levels = c("Tanycytes","Astrocytes","Ependymocytes","Neurons","Vasculars","Myelinating-glial","Oligodendrocytes","Immunes"))

obj$ann2 <- factor(obj$ann2, levels = c("Tanycytes", "Astrocytes","Ependymocytes","GABAergic","Glutamatergic","Endothelial","Pericytes","VLMCs", "Myelinating-glial", "Oligo-precursors","Oligodendrocytes","Microglia","Myeloid"))


obj$ann3 <- factor(obj$ann3, levels = c("α1-tanycytes","α2-β1-tanycytes","β2-tanycytes", "Astrocytes","Ependymocytes","GABAergic","Glutamatergic","Endothelial","Fenestrated-EC","Pericytes","VLMCs", "Myelinating-glial", "Oligo-precursors","Oligodendrocytes","Microglia","Macrophages"))

tany <- c("α1-tanycytes","α2-β1-tanycytes","β2-tanycytes")
astro <- c("Astrocytes")
oligo <- c("Oligo-precursors","Oligodendrocytes")
immune <- c("Microglia","Macrophages")
Vascular <- c("Endothelial","Fenestrated-EC","Pericytes","VLMCs")
ependy <- "Ependymocytes"
myelin <- "Myelinating-glial"

Idents(obj) <- "ann2"
sub <- subset(obj, idents = "GABAergic")
gaba <- sort(unique(sub$ann4))

Idents(obj) <- "ann2"
sub <- subset(obj, idents = "Glutamatergic")
glut <- sort(unique(sub$ann4))
rm(sub)

obj@meta.data$ann4 <- factor(obj@meta.data$ann4, levels = c(tany, astro, ependy, gaba,glut,Vascular,myelin,oligo,immune))

saveRDS(obj, file.path(dir.main, "output/000-sauve-2025/human_mediane_eminence.rds"))

```



## Analysis

```{r }

dim_plot(obj, group.by = "ann1", label = T, label.size = 3.5, legend = T, dark = T, fig.plot = T)
dim_plot(obj, group.by = "ann2", label = T, label.size = 3.5, legend = T, dark = T, fig.plot = T)
dim_plot(obj, group.by = "ann3", label = T, label.size = 3.5, legend = T, dark = T, fig.plot = T)
dim_plot(obj, group.by = "ann4", label = T, label.size = 3.5, legend = T, dark = T, fig.plot = T)
dim_plot(obj, group.by = "sample", label = F, legend = T, dark = T, fig.plot = T)
dim_plot(obj, group.by = "group", label = F, legend = T, dark = T, fig.plot = T)
dim_plot(obj, group.by = "sex", label = F, legend = T, dark = T, fig.plot = T)

```


# Helper functions

```{r}

Spatial_Feature_Plot <- function(
    seurat_object,
    features,
    images = c("slice3A"),
    image.alpha = 0,
    pt.size.factor = 50,
    interactive = FALSE,
    alpha = 1,
   ncol = NULL,
    ...
) {

  require(ggplot2)  
  require(Seurat)

  # Create plot list
  plot_list <- SpatialFeaturePlot(object = seurat_object,
                                  features = features,
                                  interactive = interactive, 
                                  pt.size.factor = pt.size.factor, 
                                  images = images, 
                                  image.alpha = image.alpha, 
                                  alpha = alpha, 
                                  ncol = ncol,
                                  ...) & theme(legend.position = "right",
                                               legend.background = element_blank(),
                                               legend.key.height = unit(0.6, 'cm'),
                                               legend.key.width = unit(0.35, 'cm'),)
  
  return(plot_list)
}




Plot_GO <- function(data,
                    fold_change, 
                    top_n = 10, 
                    specific_go_terms = NULL, 
                    plot_type = "dotplot",
                    theme_type = "minimal",
                    cluster_genes = FALSE,
                    diagonal_order = FALSE,
                    plot_title = "GO Terms Enrichment", 
                    x_title = "Genes", 
                    y_title = "GO Terms",
                    x_angle = 90, 
                    font_size = 10, 
                    wrap_length = 50,
                    alpha_value = 0.8) { 
  
  require(ggplot2)
  require(dplyr)
  require(tidyr)
  require(stringr)
  require(RColorBrewer)
  require(reshape2)
  
  colors <- rev(RColorBrewer::brewer.pal(n = 11, name = "RdBu"))
  colors <- grDevices::colorRampPalette(colors = colors)(100)
  
  data <- data.frame(data)
  data <- na.omit(data)
  data$Description <- stringr::str_to_upper(data$Description)
  
  if (!is.null(specific_go_terms)) {
    data <- data[data$Description %in% specific_go_terms, ]
    if (nrow(data) == 0) stop("No matching GO terms found.")
  } else {
    data <- data %>% arrange(p.adjust) %>% head(top_n)
  }
  
  plot_data <- data %>%
    dplyr::select(Description, geneID, p.adjust) %>%
    dplyr::mutate(Log10_padj = -log10(p.adjust)) %>%
    separate_rows(geneID, sep = "/") %>%
    dplyr::rename(GO_Term = Description, Gene = geneID)
  
  plot_data$GO_Term <- stringr::str_wrap(plot_data$GO_Term, width = wrap_length)
  
  fold_change_df <- data.frame(Gene = names(fold_change), FoldChange = fold_change)
  plot_data <- merge(plot_data, fold_change_df, by = "Gene", all.x = TRUE)
  plot_data <- plot_data %>% filter(!is.na(FoldChange))

  if (cluster_genes) {
    gene_order <- plot_data %>%
      dplyr::group_by(Gene) %>%
      dplyr::summarize(order_score = min(as.numeric(factor(GO_Term))), .groups = "drop") %>%
      dplyr::arrange(order_score)
    plot_data$Gene <- factor(plot_data$Gene, levels = gene_order$Gene)
  } else if (diagonal_order) {
    plot_data$GO_Term <- factor(plot_data$GO_Term, levels = rev(unique(plot_data$GO_Term)))
    gene_order_by_go <- plot_data %>% arrange(GO_Term) %>% pull(Gene) %>% unique()
    plot_data$Gene <- factor(plot_data$Gene, levels = gene_order_by_go)
  }

  if (plot_type == "dotplot") {
    p <- ggplot(plot_data, aes(x = Gene, y = GO_Term)) +
      geom_point(aes(size = Log10_padj, fill = FoldChange),
                 shape = 21, color = "black", alpha = alpha_value) +
      scale_size(range = c(1, 4), name = "-Log10(padj)") +
      scale_fill_gradientn(
        colors = colors,
        limits = c(min(fold_change, na.rm = TRUE), max(fold_change, na.rm = TRUE)),
        name = "LogFC",
        guide = guide_colorbar(
          barwidth = 0.7,
          barheight = 5,
          frame.colour = "black",
          ticks.colour = "black"
        )
      ) +
      plot_theme(theme_type = theme_type, x_angle = x_angle, font_size = font_size) +
      labs(title = plot_title, x = x_title, y = y_title)

  } else if (plot_type == "heatmap") {
    heatmap_matrix <- reshape2::dcast(plot_data, GO_Term ~ Gene, value.var = "FoldChange", fill = NA)
    rownames(heatmap_matrix) <- heatmap_matrix$GO_Term
    heatmap_matrix <- heatmap_matrix[, -1]
    heatmap_long <- reshape2::melt(as.matrix(heatmap_matrix), varnames = c("GO_Term", "Gene"), value.name = "FoldChange")

    if (cluster_genes) {
      heatmap_long$Gene <- factor(heatmap_long$Gene, levels = gene_order$Gene)
    } else if (diagonal_order) {
      heatmap_long$GO_Term <- factor(heatmap_long$GO_Term, levels = rev(unique(heatmap_long$GO_Term)))
      gene_order_by_go <- heatmap_long %>% arrange(GO_Term) %>% pull(Gene) %>% unique()
      heatmap_long$Gene <- factor(heatmap_long$Gene, levels = gene_order_by_go)
    }

    p <- ggplot(heatmap_long, aes(x = Gene, y = GO_Term, fill = FoldChange)) +
      geom_tile(color = "white", alpha = alpha_value) +
      scale_fill_gradientn(
        colors = colors,
        limits = c(min(fold_change, na.rm = TRUE), max(fold_change, na.rm = TRUE)),
        na.value = "white",
        name = "LogFC",
        guide = guide_colorbar(
          barwidth = 0.7,
          barheight = 5,
          frame.colour = "black",
          ticks.colour = "black"
        )
      ) +
      plot_theme(theme_type = theme_type, x_angle = x_angle, font_size = font_size) +
      labs(title = plot_title, x = x_title, y = y_title)
  } else {
    stop("Invalid plot_type. Choose 'dotplot' or 'heatmap'.")
  }
  
  return(p)
}




```



# Define colors

```{r fig.width=12, fig.height=8}

custom_colors <- list()
custom_colors$sample = c("AD-29405" = "#740001","AD-23425" = "#ae0001", "CTRL-26450" = "#6497b1", "CTRL-24457" = "#adcbe3")
custom_colors$group = c("AD" = "#740001","CTRL" = "#6497b1")

idents <- levels(obj@meta.data[["ann1"]])
colors <- colorize(x = idents, theme = "tableau", n_colors = length(idents))
custom_colors$ann1 <- colors

idents <- levels(obj@meta.data[["ann2"]])
colors <- colorize(x = idents, theme = "tableau", n_colors = length(idents))
custom_colors$ann2 <- colors

# Idents(obj) <- "ann3"
# idents <- levels(obj@meta.data[["ann3"]])
# colors <- custom_palette(n = length(idents))
# names(colors) <- idents
# custom_colors$ann3 <- colors

colors <- c(
  "Oligo-precursors"             = "#1F77B4",  # deep blue
  "GABAergic"        = "#D62728",  # clear red
  "Astrocytes"       = "#2CA02C",  # green
  "Oligodendrocytes"    = "#9467BD",  # purple
  "Glutamatergic"    = "#FF7F0E",  # orange
  "Microglia"        = "#8C564B",  # brown
  "α1-tanycytes"     = "#17BECF",  # cyan
  "Ependymocytes"    = "#BCBD22",  # olive yellow
  "α2-β1-tanycytes"  = "#1B6A6A",  # mid gray
  "VLMCs"            = "#E377C2",  # pink/magenta
  "Macrophages"      = "#B15928",  # bronze
  "β2-tanycytes"     = "#AEC7E8",  # pale blue
  "Endothelial"      = "#364A8A",  # pale green
  "Pericytes"        = "#C49C94",  # warm beige
  "Fenestrated-EC"   = "#F7B6D2",  # soft rose
  "Myelinating-glial" = "#C5B0D5"   # lavender
)

custom_colors$ann3 <- colors
  
Idents(obj) <- "ann4"
idents <- levels(obj@meta.data[["ann4"]])
colors <- custom_palette(n = length(idents))
names(colors) <- idents
custom_colors$ann4 <- colors

obj@misc$colors <- custom_colors

levels(obj$ann4)
```


# Figure 7B

```{r fig.width=7, fig.height=6}

p <- dim_plot(object = obj, group.by = "ann3", legend.title = "Cell types", label = T, legend = T, item.size = 3,  fig.plot = T, label.size = 3, colors = custom_colors$ann3, theme = "classic", plot.title = "", font.size = 10,
              legend.size = 10)
p
save_fig(fig = p, filename = file.path(dir.results, "fig7a-v1"), width = 7, height = 5)


p <- dim_plot(object = obj, group.by = "ann3", legend.title = "Cell types", label = F, legend = T, item.size = 3,  fig.plot = T, label.size = 3, colors = custom_colors$ann3, theme = "classic", plot.title = "", font.size = 10,
              legend.size = 10, n.cells = F)
p
save_fig(fig = p, filename = file.path(dir.results, "fig7a-v2"), width = 7, height = 5)


p <- dim_plot(object = obj, group.by = "ann3", legend.title = "Cell types", label = T, legend = F, item.size = 3,  fig.plot = T, label.size = 3, colors = custom_colors$ann3, theme = "classic", plot.title = "", font.size = 10,
              legend.size = 10, n.cells = F)
p
save_fig(fig = p, filename = file.path(dir.results, "fig7a-v3"), width = 6, height = 5)


```


# Fig7b

```{r}

p <- dim_plot(object = obj, group.by = "sample", legend.title = "Samples", label = F, legend = T, item.size = 3,
              fig.plot = T, label.size = 3, colors = custom_colors$sample, theme = "classic", plot.title = "",  font.size = 10, legend.size = 10)
p
save_fig(fig = p, filename = file.path(dir.results, "fig7b"), width = 6.5, height = 5)

```

# Fig7c




```{r fig.width=15, fig.height=4.5}

extract_specific_markers <- function(markers_df,
                                     pval_col = "p_val_adj",
                                     pval_thresh = 0.05,
                                     gene_col = "gene",
                                     cluster_col = "cluster",
                                     idents = NULL) {
  
  # Keep only significant genes
  sig_genes <- markers_df %>%
    dplyr::filter(.data[[pval_col]] <= pval_thresh)
  
  # Identify genes that appear in only one cluster
  specific_genes <- sig_genes %>%
    dplyr::group_by(.data[[gene_col]]) %>%
    dplyr::summarise(n_clusters = dplyr::n_distinct(.data[[cluster_col]]), .groups = "drop") %>%
    dplyr::filter(n_clusters == 1) %>%
    dplyr::pull(.data[[gene_col]])
  
  # Subset original markers to include only those specific genes
  specific_markers <- sig_genes %>%
    dplyr::filter(.data[[gene_col]] %in% specific_genes) %>%
    dplyr::distinct(.data[[gene_col]], .keep_all = TRUE)
  
  # Optionally filter by specified clusters
  if (!is.null(idents)) {
    specific_markers <- specific_markers %>%
      dplyr::filter(.data[[cluster_col]] %in% idents)
  }
  
  return(specific_markers)
}


obj <- filter_genes(obj, normalize = T, scale.data = T)

markers <- find_markers(obj,  group.by = "ann3", only.pos = T, latent.vars = c("nCount_RNA", "percent_mito"),
                        max.cells.per.ident = 500, min.pct = 0.15, logfc.threshold = 0.25, test.use = "MAST")

write.table(markers, file.path(dir.results, "ann3_markers.tsv"), sep = "\t", row.names = F, quote = F)
markers <- read.delim(file.path(dir.results, "ann3_markers.tsv"))
#top <- markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
top <- extract_specific_markers(markers_df = markers, pval_col = "p_val_adj", pval_thresh = 0.05, gene_col = "gene", cluster_col = "cluster")
top <- top %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
features = top$gene
features <- features[ !features %in% c("IL6","TRDN","GINS3") ]
features = c("SLC14A2", "CRYM", "EPHB1", "COL25A1","GPC3","FRZB", "DIO2", features)
features <- features[!features %in% c("CRYM","COL25A1","FRZB")]
features <- append(features, c("CRYM","FRZB","COL25A1"), after = 11)

p <- dot_plot(obj, features = features, group.by = "ann3", flip.axes = F, x.angle = 90, font.size = 12, theme = "test", dot.scale = 7) 
p
save_fig(fig = p, filename = file.path(dir.results, "fig7c"), width = 15, height = 4.5)

unique(markers$cluster)

```


```{r fig.width=14, fig.height=4.5}

extra_markers <- list(
  "α1-tanycytes" = c("RAX", "COL23A1", "HAS2", "DIO2", "GLUL"),
  "α2-β1-tanycytes" = c("RAX", "PAX6", "CDH2", "VIM", "FABP7","CRYM","FRZB"),
  "β2-tanycytes" = c("RAX","GPC3","COL25A1", "DIO2", "PTPRZ1", "GFAP", "CRYM"),
  "Astrocytes" = c("AQP4", "GFAP", "ALDH1L1", "SLC1A3", "GJA1"),
  "Ependymocytes" = c("FOXJ1", "S100B", "TTR", "DNAH9", "RSPH1"),
  "GABAergic" = c("GAD1", "GAD2", "SLC32A1", "DLX1", "DLX2"),
  "Glutamatergic" = c("SLC17A6", "SLC17A7", "GRIN2B", "GRIA2", "TBR1"),
  "Endothelial" = c("CLDN5", "CDH5", "PECAM1", "KDR", "FLT1"),
  "Fenestrated-EC" = c("PLVAP", "ESM1", "RGCC", "ENG", "CD93"),
  "Pericytes" = c("PDGFRB", "RGS5", "CSPG4", "MCAM", "KCNJ8"),
  "VLMCs" = c("COL1A1", "COL1A2", "PDGFRA", "DCN", "LUM"),
  "Myelinating-glial" = c("MBP", "MOG", "PLP1", "OLIG2", "MAG"),
  "Oligo-precursors" = c("PDGFRA", "CSPG4", "SOX10", "OLIG1", "VCAN"),
  "Oligodendrocytes" = c("PLP1", "MBP", "MOG", "SOX10", "MAG"),
  "Microglia" = c("CX3CR1", "P2RY12", "TMEM119", "CSF1R", "AIF1"),
  "Macrophages" = c("CD163", "MRC1", "LYVE1", "CSF1R", "S100A8")
)

```








# Fig7d

```{r fig.width=4, fig.height=4}

sub <- seurat_subset(obj,  group.by = "ann1", use.harmony = T, 
                     harmony.group = "sample", min.dist = 0.5, spread = 1, idents = c("Tanycytes"))

p <- dim_plot(object = sub, group.by = "ann3", legend.title = "", label = T, legend = F,  fig.plot = T, label.size = 5, repel = T, 
              colors = custom_colors$ann3, theme = "classic", plot.title = "", font.size = 12,legend.size = 10, pt.size = 1)
p
save_fig(fig = p, filename = file.path(dir.results, "fig7d1"), width = 4, height = 4)

```


```{r fig.width=8, fig.height=5}

p1 <- feature_plot(sub, features = c("DIO2","EPHB1", "HAS2", "CRYM","FRZB","GPC3", "COL25A1"),
                   ncol = 4, order = T, no_axes = T, merge_legend = T, base_size = 12)
p1
save_fig(fig = p1, filename = file.path(dir.results, "fig7d2"), width = 8, height = 5)

```


```{r}

## Human 10x visium from visium 2 (John A. Tadross et al, 2025)
visium <- readRDS("~/projects/data/human_HYPOMAP_spatial.rds")
visium <- UpdateSeuratObject(visium)

```

```{r fig.width=10, fig.height=8}

slice3A <- subset(visium, cells = colnames(visium@images[["slice3A"]]))
slice3A@images <- slice3A@images["slice3A"]

slice3A@meta.data$regional_clusters <- NULL
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("ME"),"regional_clusters"] <- "ME"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("ARC"),"regional_clusters"] <-"ARC"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("VMH"),"regional_clusters"] <-"VMH"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("LH/RCN"),"regional_clusters"] <-"LH"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("LTN"),"regional_clusters"] <-"LTN"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("OT"),"regional_clusters"] <-"OT"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("TMN"),"regional_clusters"] <-"."
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("LH 1"),"regional_clusters"] <-"LH"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("Fx/LH"),"regional_clusters"] <-"Fx"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("Vasculars"),"regional_clusters"] <-"Vasculars"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("Periventricular"),"regional_clusters"] <-"Periventricular"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("PVN"),"regional_clusters"] <-"."
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("Ventricular"),"regional_clusters"] <-"Ependymocytes"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("SON"),"regional_clusters"] <-"SON"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("Fx/OT/ac"),"regional_clusters"] <-"Fx"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("Unassigned"),"regional_clusters"] <-"."
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("DMH"),"regional_clusters"] <-"LH"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("LH 3"),"regional_clusters"] <-"LH"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("Thalamus 1"),"regional_clusters"] <-"."
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("Thalamus 2"),"regional_clusters"] <-"."
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("LH 2"),"regional_clusters"] <-"LH"
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("MPOA"),"regional_clusters"] <-"."
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("SCN"),"regional_clusters"] <-"."
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("LPOA"),"regional_clusters"] <-"."
slice3A@meta.data[slice3A@meta.data$regional_clusters_named %in% c("MAM 1"),"regional_clusters"] <-"DMH"

#saveRDS(visium, "~/Documents/projects/data/human_HYPOMAP_spatial.rds")

```

```{r fig.width=4, fig.height=8}

Idents(slice3A) <- "regional_clusters"
idents <- unique(slice3A@meta.data[["regional_clusters"]])
visium_colors <- custom_palette(n = length(idents))
names(visium_colors) <- idents

p <- SpatialDimPlot(slice3A, group.by = "regional_clusters", images = "slice3A", label = T, label.box = T, cols = 
                    label.color = "white", label.size = 4.5, repel = T, pt.size.factor = 100, image.alpha = 0, alpha = 10) + 
   ggtitle("Slice3A") + NoLegend()

```
```{r fig.width=20, fig.height=3}

p2 <- Spatial_Feature_Plot(slice3A, 
                     features = c("DIO2","EPHB1", "HAS2", "CRYM","FRZB","GPC3", "COL25A1"), 
                     interactive = F, 
                     pt.size.factor =200, 
                     images = "slice3A", 
                     image.alpha = 0, 
                     alpha = 1, 
                     ncol = 4)
p2
save_fig(fig = p2, filename = file.path(dir.results, "fig7d4"), width = 10, height = 10)

```




# Fig7e

```{r fig.width=6, fig.height=3.5}

p <- plot_props(obj, x = "ann3",y = "sample", plot.type  = "bar", legend.title = "",x.angle = 45,legend = T,
                stack = F, prop = T, theme = "test", colors = custom_colors$sample, x.title = "", x.lab = T, 
                legend_text_size = 10, legend_ncol = 1, legend.pos = c(0.1, 0.80), font.size = 12)
p
save_fig(fig = p, filename = file.path(dir.results, "fig7e"), width = 6, height = 3.5)

```



# Fig7f

```{r fig.width=4.5, fig.height=5}

Idents(obj) <- "ann3"
prop_test <- scProportionTest::sc_utils(obj)

prop_test <- scProportionTest::permutation_test(
	prop_test, 
	cluster_identity = "ann3",
	sample_1 = "CTRL", 
	sample_2 = "AD",
	sample_identity = "group",
	n_permutations = 10000
)

write.table(prop_test@results$permutation, file.path(dir.results, "ann3_permutation_test.tsv"),
            sep = "\t", row.names = F, quote = F)

#' Plot permutation test results for single-cell analysis
#'
#' This function creates a plot of observed log2 fold differences (log2FD) 
#' from a permutation test, highlighting statistically significant results.
#'
#' @param sc_utils_obj An object containing single-cell analysis results, 
#'   with a `@results$permutation` slot storing a `data.table` that includes:
#'   `FDR`, `obs_log2FD`, `boot_CI_2.5`, `boot_CI_97.5`, and `clusters`.
#' @param FDR_threshold Numeric, false discovery rate threshold for significance. Default is `0.05`.
#' @param log2FD_threshold Numeric, log2 fold-difference threshold for significance. 
#'   Default is `log2(1.5)`.
#' @param order_clusters Logical, if `TRUE` (default) orders clusters by observed log2FD.
#' @param theme.type Character string indicating the ggplot2 theme type to apply 
#'   via `plot_theme()`. Default is `"minimal"`.
#' @param font.size Numeric, base font size for the plot theme. Default is `10`.
#' @param legend.position Position of the legend (`"right"`, `"bottom"`, `"top"`, `"left"`). 
#'   Default is `"right"`.
#' @param legend.direction Direction of the legend (`"vertical"` or `"horizontal"`). 
#'   Default is `"vertical"`.
#' @param ... Additional arguments passed to `plot_theme()`.
#'
#' @return A `ggplot` object showing permutation results with confidence intervals 
#'   and significance highlighting.
#'
#' @details
#' The function:
#' 1. Copies the permutation results from the `sc_utils_obj` object.
#' 2. Marks results as significant if they pass both the FDR and log2FD thresholds.
#' 3. Optionally orders clusters by observed log2FD.
#' 4. Creates a point-range plot with bootstrapped confidence intervals and 
#'    color-coded significance.
#'
#' @examples
#' \dontrun{
#' p <- permut_plot(my_sc_utils_object, FDR_threshold = 0.01, log2FD_threshold = log2(2))
#' print(p)
#' }
#'
#' @export
permut_plot <- function(
    sc_utils_obj,
    FDR_threshold = 0.05,
    log2FD_threshold = log2(1.5),
    order_clusters = TRUE,
    theme = "minimal",
    colors = c("salmon", "grey60"),
    font.size = 10,
    legend.text.size = 10,
    legend.position = "right",
    legend.direction = "vertical",
    ...
) {
  ## Retrieve results.
  plot_data <- data.table::copy(sc_utils_obj@results$permutation)
  
  ## Mark the significant results.
  plot_data[, significance := ifelse(
    FDR < FDR_threshold & abs(obs_log2FD) > log2FD_threshold,
    paste("FDR <", FDR_threshold, "& abs(Log2FD) >", round(log2FD_threshold, 2)),
    "n.s."
  )]
  
  plot_data[, significance := factor(significance, levels = c(
    paste("FDR <", FDR_threshold, "& abs(Log2FD) >", round(log2FD_threshold, 2)),
    "n.s."
  ))]
  
  ## Order the clusters by observed log2FD if requested.
  if (order_clusters) {
    plot_data[, clusters := forcats::fct_reorder(factor(clusters), dplyr::desc(obs_log2FD))]
  }
  
  ## Plot the results.
  p <- ggplot(plot_data, aes(x = clusters, y = obs_log2FD)) +
    geom_pointrange(aes(ymin = boot_CI_2.5, ymax = boot_CI_97.5, color = significance)) +
    plot_theme(
      theme = theme,
      font.size = font.size,
      legend.position = legend.position,
      legend.direction = legend.direction,
      legend.text.size = legend.text.size,
      axis.title.y.show = F,
      ...
    ) +
    geom_hline(yintercept = log2FD_threshold, lty = 2) +
    geom_hline(yintercept = -log2FD_threshold, lty = 2) +
    geom_hline(yintercept = 0) +
    scale_color_manual(values = colors) +
    coord_flip()
  
  return(p)
}



p <- permut_plot(prop_test, 
                 theme = "bw", 
                 legend.position = "bottom", 
                 order.clusters = T, plot_title = "",
                 threshold.mode = "fixed",
                 font.size = 12, 
                 legend_text_size = 10, show_subtitle = F)
p
save_fig(fig = p, filename = file.path(dir.results, "fig7f"), width = 4.5, height = 5)

```


# Fig8a


```{r}

#' Generate Pseudo-Replicates from a Seurat Object by Bootstrap or Splitting
#'
#' Creates pseudo-replicates from single-cell RNA-seq data stored in a Seurat object,
#' using one of two statistical approaches:
#'
#' 1. **Bootstrap sampling ("bootstrap" method):**
#'    This method performs sampling **with replacement** of cells within each sample group.
#'    Each pseudo-replicate contains the same number of cells as the original sample,
#'    but some cells may be repeated and others omitted due to resampling.
#'    Bootstrapping approximates the sampling distribution of the data, allowing
#'    assessment of variability and robustness of downstream analyses.
#'    Because it creates a new count matrix by resampling cells, this method returns
#'    a new Seurat object with freshly sampled counts and metadata.
#'
#' 2. **Splitting ("split" method):**
#'    This method partitions the cells in each sample group into roughly equal,
#'    **non-overlapping subsets** (pseudo-replicates) by randomly splitting cells.
#'    It does not change count data or embeddings but simply annotates cells in metadata.
#'    Statistically, splitting provides multiple independent subsets for analysis,
#'    allowing evaluation of consistency or heterogeneity across subsets.
#'    This method is computationally lighter and preserves original embeddings (e.g., UMAP),
#'    since the counts are not altered.
#'
#' @param seurat_obj A Seurat object containing single-cell RNA-seq data.
#' @param sample_column Character string naming the metadata column defining groups (e.g., "donor").
#' @param target_samples Character vector of group names to generate pseudo-replicates for. Defaults to all groups.
#' @param n_replicates Integer number of pseudo-replicates to generate per group.
#' @param method Character specifying the approach to create pseudo-replicates. Options:
#'   - `"bootstrap"`: Generate bootstrap samples with replacement, returns new Seurat object.
#'   - `"split"`: Partition cells into non-overlapping groups, updates metadata only.
#' @param seed Integer random seed for reproducibility.
#'
#' @return A Seurat object:
#' - If `method = "split"`, returns the input object with updated metadata column `pseudo_replicate`.
#' - If `method = "bootstrap"`, returns a new Seurat object containing bootstrap-sampled counts and metadata.
#'
#' @examples
#' # Split method - preserves embeddings, annotates metadata - Assess consistency across independent subsets.
#' obj_split <- generate_pseudo_replicates(obj, sample_column = "sample", n_replicates = 3, method = "split")
#'
#' # Bootstrap method - new object with resampled counts - Estimate variability & robustness of DE analysis and Perform rigorous statistical inference.
#' obj_boot <- generate_pseudo_replicates(obj, sample_column = "sample", n_replicates = 3, method = "bootstrap")
#'
#' Additional notes:
#'	•	Bootstrap is generally more statistically sound for quantifying uncertainty in DE but more computationally expensive.
#'  •	Split is often used as a quick diagnostic to ensure signals are not driven by a subset of cells.
#'  •	In practice, many analyses combine both: use split to identify consistency, then bootstrap for inference.
#' @export
generate_pseudo_replicates <- function(seurat_obj,
                                       sample_column = "donor",
                                       target_samples = NULL,
                                       n_replicates = 3,
                                       method = c("split", "bootstrap"),
                                       seed = 123) {
  library(Seurat)
  set.seed(seed)
  
  method <- match.arg(method)
  
  meta <- seurat_obj@meta.data
  if (!(sample_column %in% colnames(meta))) stop(paste("Column", sample_column, "not found in metadata"))
  meta[[sample_column]] <- as.character(meta[[sample_column]])
  
  if (is.null(target_samples)) target_samples <- unique(meta[[sample_column]])
  
  if (method == "split") {
    meta$pseudo_replicate <- meta[[sample_column]]  # initialize
    
    for (sample in target_samples) {
      sel <- which(meta[[sample_column]] == sample)
      n_cells <- length(sel)
      
      if (n_cells < n_replicates) {
        warning(sprintf("Sample %s has fewer cells (%d) than requested replicates (%d). Adjusting to %d replicates.",
                        sample, n_cells, n_replicates, n_cells))
        actual_reps <- n_cells
      } else {
        actual_reps <- n_replicates
      }
      
      shuffled_cells <- sample(sel)
      split_indices <- split(shuffled_cells, rep(1:actual_reps, length.out = length(shuffled_cells)))
      
      for (rep in seq_len(actual_reps)) {
        meta$pseudo_replicate[split_indices[[rep]]] <- paste0(sample, "_rep_", rep)
      }
    }
    
    seurat_obj@meta.data <- meta
    return(seurat_obj)
  }
  
  if (method == "bootstrap") {
    counts_data <- GetAssayData(seurat_obj, assay = "RNA", slot = "counts")
    all_cells_in_counts <- colnames(counts_data)
    
    all_cells <- list()
    all_meta <- list()
    
    for (sample in target_samples) {
      sel <- rownames(meta)[meta[[sample_column]] == sample]
      sel <- intersect(sel, all_cells_in_counts)
      n_cells <- length(sel)
      if (n_cells == 0) next
      
      for (rep in seq_len(n_replicates)) {
        boot_idx <- sample(sel, size = n_cells, replace = TRUE)
        rep_label <- paste0(sample, "_boot_", rep)
        
        counts_mat <- counts_data[, boot_idx, drop = FALSE]
        colnames(counts_mat) <- paste0(rep_label, "_", seq_along(boot_idx))
        
        meta_rep <- meta[boot_idx, , drop = FALSE]
        rownames(meta_rep) <- colnames(counts_mat)
        meta_rep$pseudo_replicate <- rep_label
        
        all_cells[[rep_label]] <- counts_mat
        all_meta[[rep_label]] <- meta_rep
      }
    }
    
    counts_combined <- do.call(cbind, all_cells)
    meta_combined <- do.call(rbind, all_meta)
    
    new_obj <- CreateSeuratObject(counts = counts_combined, meta.data = meta_combined)
    return(new_obj)
  }
}


```

```{r}


# DE pseudobulk

## Generate pseudo-replicates

obj <- generate_pseudo_replicates(obj, sample_column = "sample", n_replicates = 3, method = "split")

table(obj@meta.data$pseudo_replicate)

dim_plot(obj, group.by = "pseudo_replicate")

plot_props(obj, x = "ann_level_3",y = "pseudo_replicate", plot_type = "bar", legend_title = "",x_angle = 45,legend = T,
                stack = T, prop = T, theme_type = "classic", x_title = "", x_lab = T, 
                legend_text_size = 8, legend_ncol = 1, font_size = 12)

```



```{r}

Idents(obj) <- "ann3"
obj@meta.data$ann3 <- as.character(obj@meta.data$ann3)

Idents(obj) <- "ann3"
sub <- subset(obj, idents = c("α1-tanycytes","α2-β1-tanycytes","β2-tanycytes"))
sub <- filter_genes(sub)
sub <- generate_pseudo_replicates(sub, sample_column = "sample", n_replicates = 3, method = "split")

pb <- DE_libra(object = sub, ident_col = "ann3", sample_col = "pseudo_replicate", 
               label_col = "group", label_levels = c("CTRL", "AD"), de_method = "edgeR", 
               de_type = "QLF",label1 = "CTRL", label2 = "AD", min_samples_gene = 6)

saveRDS(pb, file.path(dir.results, "EdgeR_QLF_AD_vs_CTRL.rds"))
write.table(pb$DE_sig_all, file.path(dir.results, "EdgeR_QLF_AD_vs_CTRL.tsv"), sep = "\t", row.names = F, quote = F)

```



```{r}

# Only keep cell types present in your DE results
pb <- readRDS(file.path(dir.results, "EdgeR_QLF_AD_vs_CTRL.rds"))
DE = pb$DE

# Remove NA cell types → this is the REAL fix
DE <- DE[!is.na(DE$cell_type), ]

# Keep only desired cell types
cell_order <- c("α1-tanycytes","α2-β1-tanycytes","β2-tanycytes","Astrocytes","Ependymocytes","GABAergic",
                "Glutamatergic","Endothelial","Fenestrated-EC","Pericytes","VLMCs","Myelinating-glial",
                "Oligo-precursors","Oligodendrocytes","Microglia","Macrophages")

cell_order_present <- cell_order[cell_order %in% DE$cell_type]

# Set factor order
DE$cell_type <- factor(DE$cell_type, levels = cell_order_present)

# Now run the plot
p <- plot_de_summary(
  DE = DE,
  cell.type.col = "cell_type",
  pvalue.col = "p_val_adj",
  logFC.col = "avg_logFC",
  pvalue = 0.05,
  logFC = 0,
  plot.title = "",
  font.size = 12,
  show.text = FALSE,
  axis.text.size = 8,
  custom.colors = c("Up"="#740001","Down"="#6497b1"),
  x.angle = 45,
  flip.coords = FALSE,
  theme = "bw",
  dodge.plot = TRUE,
  legend.position = "top",
  legend.direction = "horizontal",
  cell.type.order = cell_order_present
)



print(p)
save_fig(fig = p, filename = file.path(dir.results, "fig8a"), width = 6, height = 4)


```





# Fig8b


```{r fig.width=}

get_top_DE_genes <- function(pb, 
                             celltype, 
                             top_n = 50, 
                             extra_genes = NULL) {
  # 1. Filter significant DE for the given cell type
  DE_sig_ct <- pb$DE_sig_all %>%
    dplyr::filter(cell_type == celltype)
  
  if (nrow(DE_sig_ct) == 0) {
    warning(paste("No DE_sig rows found for cell type:", celltype))
    return(NULL)
  }
  
  # 2. Select top_n genes by absolute logFC
  top_genes <- DE_sig_ct %>%
    dplyr::slice_max(order_by = abs(avg_logFC), n = top_n, with_ties = FALSE) %>%
    dplyr::pull(gene)
  
  # 3. Combine with extra genes
  deg_genes <- union(top_genes, extra_genes)
  
  # 4. Subset the *full* DE table for these genes
  DE_subset <- pb$DE_sig_all %>%
    dplyr::filter(cell_type == celltype, gene %in% deg_genes) %>%
    dplyr::arrange(desc(abs(avg_logFC)))
  
  return(DE_subset)
}

genes <- c("SPP1", "HIF1A", "HIF3A", "VEGFA", "ENO1","MT1X", "MT1M", "MT1E", "MT2A", "MT3","SOD2", "SOD3",
                  "SESN1", "SESN2","SELENOS", "GPX3","BNIP3", "DDIT4","CLU", "CSF", "C3")

# obj <- generate_pseudo_replicates(obj, sample_column = "sample",n_replicates = 2, method = "bootstrap")
# obj$pseudo_replicate = factor(obj$pseudo_replicate)

idents <- unique(obj@meta.data[["pseudo_replicate"]])
pseudo_colors <- colorize(x = idents, theme = "RdYlBu", n_colors = length(idents))

# Fig8B

deg <- get_top_DE_genes(pb = pb, celltype = "α1-tanycytes", top_n = 30,extra_genes = genes)

Plot_DE_Pseudobulk(DE = deg , matrices = pb$matrices, n_lab = length(deg$gene), anno_legend = F, 
                   border_color = "grey40", group1_pattern = "CTRL", group2_pattern = "AD", 
                   plot_type = "both",color_theme = "viridis",
                   custom_annotation_colors = list(Samples = pseudo_colors, Groups = custom_colors$group),
                   legend = T,show.rownames = T, show.colnames = F, width = 5, height = 2, min_samples_gene = 2,
                   base_size = 14, out_dir = file.path(dir.results, "heatmap"), dynamic_height = 10, highlight_genes = genes)

# Fig8C

deg <- get_top_DE_genes(pb = pb, celltype = "α2-β1-tanycytes", top_n = 30, extra_genes = genes)
dim(deg)

Plot_DE_Pseudobulk(DE = deg, matrices = pb$matrices, n_lab = length(deg$gene), anno_legend = F, border_color = "grey40",
                   group1_pattern = "CTRL", group2_pattern = "AD", plot_type = "heatmap",color_theme = "viridis",
                   custom_annotation_colors = list(Samples = pseudo_colors, Groups = custom_colors$group),
                   legend = T,show.rownames = T, show.colnames = F, width = 5, height = 2, min_samples_gene = 2,
                   base_size = 14, out_dir = file.path(dir.results, "heatmap"), dynamic_height = 10, highlight_genes = genes)


# Fig8D

deg <- get_top_DE_genes(pb = pb, celltype = "β2-tanycytes", top_n = 40, extra_genes = genes)

Plot_DE_Pseudobulk(DE = deg, matrices = pb$matrices, n_lab = length(deg$gene), anno_legend = T, border_color = "grey40",
                   group1_pattern = "CTRL", group2_pattern = "AD", plot_type = "heatmap",color_theme = "viridis",
                   custom_annotation_colors = list(Samples = pseudo_colors, Groups = custom_colors$group),
                   legend = T,show.rownames = T, show.colnames = F, width = 7, height = 2, min_samples_gene = 2,
                   base_size = 14, out_dir = file.path(dir.results, "heatmap"), dynamic_height = 10, highlight_genes = genes)


```








```{r fig.width=12, fig.height=4}


DE_tany <- subset(res$DE_global, celltype == c("α1-tanycytes", "α2-β1-tanycytes", "β2-tanycytes"))
unique(DE_tany$celltype)
enhanced_volcano(df = DE_tany, celltype_col = "celltype", pval_col = "p_val", logFC_cutoff = 0.1, 
                 comparison_col = "comparison", highlight_genes = genes, ncol = 3)


highlight_genes <- list(
  "α1-tanycytes" = c("PDE1C","ANKRD20A4P","FGD3","ARL17B","ZPBP","TRPV2","EPSTI1","GALNT17","ADAMTS12",
"ARHGAP25","SNX31","ID1","IGFN1","DCC","NRXN3","DDIT4","VEGFA","BNIP3","AVP","ASNS","HILPDA","ADM",
"SOD3","PLXNA4","PGAP1","ATP1B3","NUP210L","HK2","CALCB","PDK1","PDE3B","IGFBP5","ST3GAL6","ENO1"),
  "α2-β1-tanycytes" = c("SESN2", "DDIT4", "MT1X", "VEGFA", "HIF1A", "PRLHR", "CHAC1", "SLC22A3",
  "HIF3A", "PDE3B", "NUP210L", "A2M", "TSPAN2", "ADCY1", "INSYN2B", "SLC28A1",
  "SLC9C1", "ARAP3", "EGR1", "RANBP17", "SPTSSB", "ZNF804A", "LRP2", "PRCD",
  "ADAMTS12", "HCN1", "GDPD5", "ARL17B", "BEST3", "HMCN1", "A2M", "MYO16",
  "SAMD5", "BTBD11", "CALCRL", "CDH18", "ADAMTS18", "COL1A2", "AK5", "PCSK2",
  "TENM2", "OPCML", "ANKRD20A4P", "SNX31", "ARHGAP25", "IRAK3", "SLITRK5",
  "ZIC1", "GRHL3", "PRR16", "SESN2", "PI15", "GPRC5A", "ROBO2", "RYR3",
  "KCNE4", "ANGPTL4", "SOD2", "VEGFA", "DDIT4", "BNIP3", "HILPDA", "SOD3",
  "AVP", "CLU", "MT1X", "SELENOS", "NIBAN1", "MT1M","ENO1"),
  "β2-tanycytes" = c("ARL17B","PDE3A","FAP","RGS4","WFIKKN2","MT1X","MT3","SELENOS","HIF3A","ZPLD1",
  "FAM237B","RALYL","A2M","KITLG","CPED1","FHL2","BMF","RRH","MYH11","PRICKLE1",
  "TC2N","AK5","FNDC1","KCNIP1","RUNX1","ENO1")
)

p <- enhanced_volcano(
  df = DE_tany,
  celltype_col = "celltype",
  pval_col = "p_val_adj",
  logFC_cutoff = 1,
  comparison_col = "comparison",
  highlight_genes = highlight_genes,
  ncol = 3
)
p
save_fig(fig = p, filename = file.path(dir.results, "fig8b"), width = 12, height = 4)


```



# Run GO BP


```{r}

EnrichGO <- function(de_table, 
                       gene_col = "gene",
                       cluster_col = "cell_ype",  
                       logFC_col = "avg_logFC",  
                       pval_col = "p_val_adj",  
                       pval_cutoff = 0.05, 
                       p_adj_method = "none",
                       ontology = "BP",  
                       min_genes = 10,  
                       output_dir = "Enrich_GO") {
  
  require(clusterProfiler)
  require(org.Hs.eg.db)
  require(dplyr)
  
  # Ensure required columns exist
  required_cols <- c(gene_col, cluster_col, logFC_col, pval_col)
  if (!all(required_cols %in% colnames(de_table))) {
    stop("Error: DE table must contain the specified columns: ", paste(required_cols, collapse=", "))
  }

  # Filter for significant genes
  de_table <- de_table %>% filter(.data[[pval_col]] < pval_cutoff)
  
  # Get unique clusters (e.g., cell types)
  unique_clusters <- unique(de_table[[cluster_col]])
  
  # Store results
  dir <- file.path(output_dir)
  if (!dir.exists(dir)) dir.create(dir, recursive = TRUE)
  
  go_results_list <- list()
  
  for (cluster in unique_clusters) {
    cat("Running GO enrichment for:", cluster,"...\n")
    
    # Separate Up and Down genes
    up_genes <- de_table %>%
      filter(.data[[cluster_col]] == cluster, .data[[logFC_col]] > 0) %>%
      pull(.data[[gene_col]])
    
    down_genes <- de_table %>%
      filter(.data[[cluster_col]] == cluster, .data[[logFC_col]] < 0) %>%
      pull(.data[[gene_col]])
    
    # Function to perform GO enrichment
    run_go <- function(gene_list, regulation_status) {
      if (length(gene_list) < min_genes) return(NULL)  # Skip if too few genes
      
      entrez_genes <- mapIds(org.Hs.eg.db, keys = gene_list, keytype = "SYMBOL", column = "ENTREZID", multiVals = "first")
      entrez_genes <- na.omit(entrez_genes)  # Remove NAs
      
      if (length(entrez_genes) >= min_genes) {
        gse <- enrichGO(gene = entrez_genes,
                              OrgDb = org.Hs.eg.db,
                              keyType = "ENTREZID",
                              ont = ontology,  
                              pAdjustMethod = p_adj_method,
                              pvalueCutoff = pval_cutoff,
                              #qvalueCutoff = pval_cutoff,
                              readable = TRUE)
        
        gse@result$Description <- stringr::str_to_upper(gse@result$Description)
        go_result <- gse@result

        if (!is.null(go_result) && nrow(go_result) > 0) {
          go_df <- as.data.frame(go_result)
          
          # Only add metadata if results exist
          if (nrow(go_df) > 0) {
            go_df[[cluster_col]] <- cluster  # Add cluster column
            go_df$Regulation <- regulation_status  # "Up" or "Down"
            saveRDS(gse, file.path(dir, paste0(cluster, "_Enrich_GO_", ontology, ".rds")))
          }
          
          return(go_df)
        }
      }
      return(NULL)
    }
    
    # Run GO for Up and Down genes
    up_results <- run_go(up_genes, "Up")
    down_results <- run_go(down_genes, "Down")
    
    # Store only non-empty results
    if (!is.null(up_results) && nrow(up_results) > 0) {
      go_results_list[[paste0(cluster, "_Up")]] <- up_results
    }
    if (!is.null(down_results) && nrow(down_results) > 0) {
      go_results_list[[paste0(cluster, "_Down")]] <- down_results
    }
  }
  

  # Merge all results
  if (length(go_results_list) > 0) {
    final_go_results <- bind_rows(go_results_list)
    write.table(final_go_results, file.path(dir, paste0("Enrich_GO_", ontology, ".tsv")),sep = "\t", row.names = F, quote = F)
    cat("GO enrichment completed. Results saved in: ", dir,"\n")
    return(final_go_results)
  } else {
    cat("No significant GO terms found for any cell type.\n")
    return(NULL)
  }
}


DE_sub <- subset(pb$DE, cell_type == c("α1-tanycytes", "α2-β1-tanycytes", "β2-tanycytes") & p_val_adj <= 0.05)

go <- EnrichGO(de_table = pb$DE, gene_col = "gene", cluster_col = "cell_type", logFC_col = "avg_logFC",
  pval_col = "p_val_adj", pval_cutoff = 0.05, p_adj_method = "none",ontology = "BP",min_genes = 5,
  output_dir = file.path(dir.results, "GO_BP")
)

go$Description <- stringr::str_to_upper(go$Description)


go_terms_of_interest <- c(
  "DETOXIFICATION",
  "CELLULAR RESPONSE TO HYPOXIA",
  "STRESS RESPONSE TO METAL ION",
  "RESPONSE TO REACTIVE OXYGEN SPECIES",
  "COMPLEMENT ACTIVATION, CLASSICAL PATHWAY",
  "BLEB ASSEMBLY",
  "LIPID TRANSPORT",
  "REGULATION OF LIPID METABOLIC PROCESS",
  "GLUCOSE HOMEOSTASIS",
  "CALCIUM ION HOMEOSTASIS",
  "POSTSYNAPSE ORGANIZATION",
  "VESICLE DOCKING INVOLVED IN EXOCYTOSIS",
  "REGULATION OF SYNAPTIC VESICLE EXOCYTOSIS",
  "THYROID HORMONE METABOLIC PROCESS",
  "THYROID HORMONE TRANSPORT",
  "ERK1 AND ERK2 CASCADE",
  "LIGAND-GATED ION CHANNEL SIGNALING PATHWAY",
  "CALCIUM ION TRANSPORT",
  "MONOATOMIC ANION TRANSPORT"
)

```


```{r}
plot_go_terms <- function(go_df,
                          go_terms_of_interest = NULL,
                          cluster,
                          top_n = 10,
                          wrap_width = 50,
                          color_values = c("Up" = "#740001", "Down" = "#6497b1"),
                          size_range = c(2, 6),
                          theme = "bw",
                          x_angle = 0,
                          font_size = 12,
                          plot_title_size = 16) {
  
  go_df$Description <- stringr::str_to_upper(go_df$Description)

  # Filter for the target cell type
  df <- go_df %>%
    dplyr::filter(.data[["cell_type"]] == cluster)
  
  # If terms are NULL, select top_n by FoldEnrichment
  if (is.null(go_terms_of_interest)) {
    go_terms_of_interest <- df %>%
      dplyr::arrange(desc(.data[["FoldEnrichment"]])) %>%
      dplyr::pull(.data[["Description"]]) %>%
      unique() %>%
      head(top_n)
  }
  
  # Filter for selected terms
  df <- df %>%
    dplyr::filter(.data[["Description"]] %in% go_terms_of_interest) %>%
    dplyr::arrange(desc(.data[["FoldEnrichment"]])) %>%
    dplyr::group_by(.data[["Description"]]) %>%
    dplyr::slice(1) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(
      log10FDR = -log10(.data[["pvalue"]]),
      Count = as.integer(.data[["Count"]]),
      CountSize = Count * 0.8,
      Regulation = factor(.data[["Regulation"]], levels = c("Up", "Down"))
    )
  
  # Wrap long GO term names
  df[["Description"]] <- stringr::str_wrap(df[["Description"]], width = wrap_width)
  
  # Plot
  p <- ggplot(df, aes(x = log10FDR, y = reorder(.data[["Description"]], log10FDR))) +
    geom_segment(aes(x = 0, xend = log10FDR,
                     yend = reorder(.data[["Description"]], log10FDR),
                     color = Regulation),
                 size = 0.8, alpha = 1) +
    geom_point(aes(size = .data[["FoldEnrichment"]], color = Regulation), alpha = 1) +
    scale_size_identity(name = "Gene Count", guide = "legend",
                        breaks = df[["FoldEnrichment"]], labels = df$Count) +
    scale_color_manual(values = color_values) +
    scale_size(range = size_range) +
    labs(
      x = expression(-log[10](FDR)),
      y = "",
      size = "Enrichment",
      color = "Regulation",
      title = cluster
    ) +
    plot_theme(theme = theme, x.angle = x_angle, font.size = font_size)
  
  return(p)
}
```


```{r}
a1 = c("cell growth",
"cellular response to hypoxia",
"neuron apoptotic process",
"reactive oxygen species metabolic process",
"response to endoplasmic reticulum stress",
"regulation of metal ion transport",
"intrinsic apoptotic signaling pathway",
"cyclic nucleotide metabolic process",
"cell junction maintenance",
"cell-matrix adhesion",
"regulation of synaptic vesicle exocytosis",
"glucose homeostasis",
"cilium assembly",
"regulation of lipid metabolic process",
"microtubule-based transport")

b2 <- c("cell growth",
"cellular response to hypoxia",
"neuron apoptotic process",
"reactive oxygen species metabolic process",
"response to endoplasmic reticulum stress",
"regulation of metal ion transport",
"stress response to metal ion",
"cyclic nucleotide metabolic process",
"cell junction maintenance",
"cell-matrix adhesion",
"regulation of synaptic vesicle exocytosis",
"glucose homeostasis",
"fatty acid metabolic process",
"regulation of actin filament-based movement make vector with all upper case")

a1 <-stringr::str_to_upper(a1)

b2 <-stringr::str_to_upper(b2)

go <- read.delim(file.path(dir.results, "GO_BP/Enrich_GO_BP.tsv"))

```


```{r fig.width=8, fig.height=5}

p <- plot_go_terms(go_df = go, go_terms_of_interest = a1, cluster = "α1-tanycytes")
p
save_fig(fig = p, filename = file.path(dir.results, "fig8c1"), width = 8, height = 5)

p <- plot_go_terms(go_df = go, go_terms_of_interest = go_terms_of_interest, cluster = "α2-β1-tanycytes")
p
save_fig(fig = p, filename = file.path(dir.results, "fig8c2"), width = 8, height = 5)

p <- plot_go_terms(go_df = go, go_terms_of_interest = b2, cluster = "β2-tanycytes")
p
save_fig(fig = p, filename = file.path(dir.results, "fig8c3"), width = 8, height = 5)

```
# Supp_Fig6b

```{r fig.width=6.5, fig.height=7}

p1 <- summarize_seurat(object = obj, 
                 group.by = "ann3", 
                 condition.by = "sex", 
                 plot.variable = "total_transcripts",
                 legend = T, theme = "classic",
                 legend.ncol = 2,
                 plot.title = "",
                 legend.title = "Gender",
                 reorder.bars = F, 
                 remove.x.labels = T,
                 remove.x.title = T,
                 remove.y.title = F,
                 legend.position = c(0.10, 0.90),
                 x.title = "celltype", 
                 y.title = "Counts", text.size = 12,
                 custom.colors = c("pink","skyblue"))

save_fig(fig = p1$plot, filename = file.path(dir.results, "supp_fig6B1"), width = 8, height = 5)


p2 <- violin_plot(obj, features = c("nFeature_RNA", "nCount_RNA", "percent_mito"), group.by = "ann3", x.angle = 45, plt.title.position = "left", 
                   pt.size = 0, font.size = 12, plot.title = "", colors = obj@misc$colors$ann3, stacked = T, ncol = 1)

save_fig(fig = p2, filename = file.path(dir.results, "supp_fig6B2"), width = 8, height = 7)

```





# Sankey plot

```{r fig.width=12, fig.height=10}

colors <- list(
  "ann1" = custom_colors$ann1,
  "ann2" = custom_colors$ann2,
  "ann3" = custom_colors$ann3,
  "ann4" = custom_colors$ann4

)
```


```{r fig.width=12, fig.height=10}

p <- sankey_plot(obj, 
              selected_vars = c("ann1", "ann2","ann3","ann4"),
              custom_colors = colors,  
              show_percentages = T,
              show_counts = F, 
              show_labels = T,
              flow.alpha = 0.5,
              label.justify = "right", 
              label.nudge = 0.5, 
              prevent_overlap = T)
print(p)
save_fig(fig = p, filename = file.path(dir.results, "supp_fig7a"), width = 12, height = 10)


```


# supp_fig7b

```{r fig.width=18, fig.height=7}

#obj <- filter_genes(obj, normalize = T, scale.data = T)
markers <- find_markers(obj,  group.by = "ann4", only.pos = T, max.cells.per.ident = 500, min.pct = 0.15, logfc.threshold = 0.25, test.use = "MAST")
write.table(markers, file.path(dir.results, "ann4_markers.tsv"), sep = "\t", row.names = F, quote = F)
markers <- read.delim(file.path(dir.results, "ann4_markers.tsv"))
top <- markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_log2FC)

p <- dot_plot(obj, features = top$gene, group.by = "ann4", flip.axes = F,  x.angle = 90, font.size = 12, theme = "test") 
p
save_fig(fig = p, filename = file.path(dir.results, "supp_fig7b"), width = 18, height = 7)

```


# Supplementary tables

```{r fig.width=10, fig.height=5}

invisible(gc(reset = T))
p <- summarize_seurat(obj, group.by = "sample", plot.variable = "num_cells", reorder.bars = T, legend = F, x.angle = 90, text.size = 8)
p
write.table(p$summary_table, file.path(dir.results, "Supplementary_Table_2.tsv"), sep = "\t", row.names = F, quote = F)

```


```{r}

df <- cluster_stats(obj, group.by = c("sex", "group","sample"), ident = "ann3",rm.sum = F, rm.pct = T)
write.table(df, file.path(dir.results, "Supplementary_Table_1.tsv"), sep = "\t", row.names = FALSE, quote = FALSE)

```

## Session info
```{r session_info}
utils::capture.output(devtools::session_info())
```


---
## Navigation
- [Next step →](001_integration.Rmd)
